### ISSUE TITLE
`list` type annotations do not allow default values and specifying generic type

### ISSUE THREAD

**Issue by amtoine** [nushell/nushell#8109]
hello there :wave: :yum: 

### Describe the bug

Related to the second part of what i'm saying in #7699, the other first part is being addressed in #7699 :ok_hand:

i'm having trouble making the type annotations of compound types work as i expect
> **Note**
> here i take the example of `list`s, which is a simple compound type

i've got two main issues, as stated in the title of this issue :yum: 
- default `list` values do not allow a function definition to compile
- i can not give the generic type of `list<T>` to a signature

### 1. the original simple error with default values
define the following function
```bash
def foo [my_list: list = ["default value"]] {
  $my_list
}
```
and get the error
```bash
Error: nu::parser::assignment_mismatch (link)

  × Default value wrong type
   ╭─[entry #30:1:1]
 1 │ def foo [my_list: list = ["default value"]] {
   ·                          ────────┬────────
   ·                                  ╰── default value not list<any>
 2 │   $my_list
   ╰────
```

#### 2.1. no default value
```bash
def foo [my_list: list = []] {
  $my_list
}
```
and
```bash
> foo [1, 2, 3]
╭───┬───╮
│ 0 │ 1 │
│ 1 │ 2 │
│ 2 │ 3 │
╰───┴───╯
```

#### 2.2. complete the annotation
```bash
def foo [my_list: list<string> = ["a default string"]] {
  $my_list
}
```
gives
```
Error: nu::parser::unknown_type (link)

  × Unknown type.
   ╭─[entry #39:1:1]
 1 │ def foo [my_list: list<string> = ["a default string"]] {
   ·                   ──────┬─────
   ·                         ╰── unknown type
 2 │   $my_list
   ╰────
```

### 1. give default values
```bash
def foo [my_list: list = ["a default string"]] {
  $my_list
}
```
or even
```bash
def foo [my_list: list<any> = ["a default string"]] {
  $my_list
}
```
would give
```bash
> foo
╭───┬──────────────────╮
│ 0 │ a default string │
╰───┴──────────────────╯
> foo ["another string"]
╭───┬────────────────╮
│ 0 │ another string │
╰───┴────────────────╯
> foo [1234]
╭───┬──────╮
│ 0 │ 1234 │
╰───┴──────╯
```

### 2. generic type annotations
we could even restrict the type with `<T>`
```bash
def foo [my_list: list<string> = ["a default string"]] {
  $my_list
}
```
would result in
```bash
> foo
╭───┬──────────────────╮
│ 0 │ a default string │
╰───┴──────────────────╯
> foo ["another string"]
╭───┬────────────────╮
│ 0 │ another string │
╰───┴────────────────╯
```
but the following would give an error like
```bash
> foo [1234]
Error: nu::parser::type_mismatch (link)

  × Type mismatch.
   ╭─[entry #15:1:1]
 1 │ foo [1234]
   ·     ───┬──
   ·        ╰── expected List(String), found List(Int)
   ╰────
```

### Configuration

| key                | value                                    |
| ------------------ | ---------------------------------------- |
| version            | 0.75.1                                   |
| branch             |                                          |
| commit_hash        | 66398fbf77d03a6c7fd231da157f410d0446132e |
| build_os           | linux-x86_64                             |
| build_target       | x86_64-unknown-linux-gnu                 |
| rust_version       | rustc 1.66.1 (90743e729 2023-01-10)      |
| rust_channel       | 1.66.1-x86_64-unknown-linux-gnu          |
| cargo_version      | cargo 1.66.1 (ad779e08b 2023-01-10)      |
| pkg_version        | 0.75.1                                   |
| build_time         | 2023-02-18 11:04:21 +01:00               |
| build_rust_channel | release                                  |
| features           | database, default, trash, which, zip     |
| installed_plugins  | gstat                                    |

**Comment by fdncred** [comment]
Have you tried out this yet to see if it helps? https://github.com/nushell/nushell/pull/8096

**Comment by amtoine** [comment]
> Have you tried out this yet to see if it helps? #8096
>
nope, but i sure can :+1:

**Comment by amtoine** [comment]
i think it works :open_mouth: 
at least for the first part of this issue :+1: 

we still can not narrow down the types of the arguments with `list<any>` or list<string>` for instance.

**Comment by fdncred** [comment]
I looked at adding `list<string>` etc, but couldn't figure it out. It'll take someone smarter with the parser than me.

**Comment by amtoine** [comment]
> I looked at adding `list<string>` etc, but couldn't figure it out. It'll take someone smarter with the parser than me.
>
no worries :+1: 

then we might want to keep that open for future tracking

**Comment by amtoine** [comment]
the first part of this issue has been fixed by #8096 :ok_hand:
